### ISSUE TITLE
Serilog.FileLoggerConfigurationExtensions.File is not backward compatible

### ISSUE THREAD

**Issue by krurkrit** [serilog/serilog-sinks-file#244]
I got the error "Method not found: Serilog.LoggerConfiguration Serilog.FileLoggerConfigurationExtensions.File" and like the issue on https://github.com/serilog/serilog-sinks-file/issues/159.

I already investigate and found the commit 1596f80d66c65ca0cb22358dbaeaa871321a7b99, the adding parameter retainedFileTimeLimit at line 259 affect method File is not backward compatible.

Cause is the "Optional parameters" not backward compatible and not support "Method overloading" because the "Optional parameters" in C# resolve the symbol at the compile time. (It's a critical recommend in SonarQube https://rules.sonarsource.com/csharp/RSPEC-2360)

Let say we have the another nuget package use "Serilog.FileLoggerConfigurationExtensions.File", I will named it "LibA".  The "LibA" add "Serilog" to build and call method "File" with 15 parameters. When nuget "LibA" run, it will expect method "File" with 15 parameters from "Serilog".

If you add an optional parameter to method "File", there are 16 parameters and nuget "LibA" cannot find it.

The "Optional parameters" in C# resolve the symbol at the compile time means.
If there are 15 parameters and 5 parameters are optional parameters.

When compile "LibA" and it call method "File" with 10 parameters, the compiler will automatically add 5 optional parameters with default value. So, when "LibA" run it expect nuget "serilog" to provide method "File" with 15 parameters.

So, the adding the optional parameter is not backward compatible.

There are 2 solutions to solve this issue.
1. Ask "LibA" contributor to add the new "Serilog" nuget and rebuild the "LibA" nuget. However, it means every nuget packages that use "Serilog" require to rebuild everytime that "Serilog" add more optional parameters.
2. "Serilog" should change to "Method overloading" for public method.

**Comment by augustoproiete** [comment]
This is by design. The [commit](https://github.com/serilog/serilog-sinks-file/commit/1596f80d66c65ca0cb22358dbaeaa871321a7b99) you referenced was released with Serilog.Sinks.File **5.0.0** which is a major release, and [major releases are allowed to introduce breaking changes to the API](https://semver.org/#semantic-versioning-200). The existence of breaking changes is the exact reason the version jumped from 4.1.0 to 5.0.0... Otherwise it would have been released as 4.2.0 to communicate backward compatibility.

If "LibA" 1.0.0 is targets Serilog.Sinks.File 4.1.0 with `X` parameters, and Serilog.Sinks.File 5.0.0 introduces new ones and now have `Y` parameters, then:

**1.** Users of `LibA` should explicitly reference the compatible version Serilog.Sinks.File 4.1.0, instead of getting the latest. e.g.

```xml
<PackageReference Include="LibA" Version="1.0.0" />
<PackageReference Include="Serilog.Sinks.File" Version="4.1.0" />
```

and/or

**2.** The developers of `LibA` should release a 2.0.0 that targets Serilog.Sinks.File 5.0.0. Users can then choose to upgrade to the new version of `LibA` and leverage the more recent version of Serilog.Sinks.File that it depends on

```xml
<PackageReference Include="LibA" Version="2.0.0" />
```

---

TL;DR; It's unreasonable to expect backward compatibility when upgrading to a major release.

**Comment by krurkrit** [comment]
@augustoproiete 

I understand that it's major release, but I think it's easy to fix by adding method overloading.
However, It's ok, I could change back to use 4.1.0 or request LibA to provide the Nuget package that build with Serilog.Sinks.File 5.0.0.

I just want to make sure that Serilog developers (or reviewers) understand the problem from the "**optional parameters**" because the optional parameter changing (add the new one or just change the default value) in minor or patch version, it might break the other Nuget packages that depends on it.

Thank you
